src/cluster/Backend.cc

```diff
@@ -0,0 +1,276 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#include "zeek/cluster/Backend.h"
+
+#include <optional>
+
+#include "zeek/Desc.h"
+#include "zeek/Event.h"
+#include "zeek/Func.h"
+#include "zeek/Reporter.h"
+#include "zeek/Type.h"
+#include "zeek/cluster/Serializer.h"
+#include "zeek/iosource/Manager.h"
+
+using namespace zeek::cluster;
+
+std::string_view detail::Event::HandlerName() const {
+    if ( std::holds_alternative<FuncValPtr>(handler) )
+        return std::get<FuncValPtr>(handler)->AsFunc()->GetName();
+
+    return std::get<EventHandlerPtr>(handler)->Name();
+}
+
+namespace {
+
+std::optional<zeek::Args> check_args(const zeek::FuncValPtr& handler, zeek::ArgsSpan args) {
+    const auto& func_type = handler->GetType<zeek::FuncType>();
+
+    if ( func_type->Flavor() != zeek::FUNC_FLAVOR_EVENT ) {
+        zeek::reporter->Error("unexpected function type for %s: %s", handler->AsFunc()->GetName().c_str(),
+                              func_type->FlavorString().c_str());
+        return std::nullopt;
+    }
+
+    const auto& types = func_type->ParamList()->GetTypes();
+    if ( args.size() != types.size() ) {
+        zeek::reporter->Error("bad number of arguments for %s: got %zu, expect %zu",
+                              handler->AsFunc()->GetName().c_str(), args.size(), types.size());
+        return std::nullopt;
+    }
+
+    zeek::Args result(args.size());
+
+    for ( size_t i = 0; i < args.size(); i++ ) {
+        const auto& a = args[i];
+        const auto& got_type = a->GetType();
+        const auto& expected_type = types[i];
+
+        if ( ! same_type(got_type, expected_type) ) {
+            zeek::reporter->Error("event parameter #%zu type mismatch, got %s, expecting %s", i,
+                                  zeek::obj_desc_short(got_type.get()).c_str(),
+                                  zeek::obj_desc_short(expected_type.get()).c_str());
+            return std::nullopt;
+        }
+
+        result[i] = args[i];
+    }
+
+    return result;
+}
+
+} // namespace
+
+std::optional<detail::Event> Backend::MakeClusterEvent(FuncValPtr handler, ArgsSpan args, double timestamp) const {
+    auto checked_args = check_args(handler, args);
+    if ( ! checked_args )
+        return std::nullopt;
+
+    if ( timestamp == 0.0 )
+        timestamp = zeek::event_mgr.CurrentEventTime();
+
+    return zeek::cluster::detail::Event{handler, std::move(*checked_args), timestamp};
+}
+
+zeek::RecordValPtr Backend::DoMakeEvent(zeek::ArgsSpan args) {
+    static const auto& any_vec_type = zeek::id::find_type<zeek::VectorType>("any_vec");
+    static const auto& event_record_type = zeek::id::find_type<zeek::RecordType>("Cluster::Event");
+    auto rec = zeek::make_intrusive<zeek::RecordVal>(event_record_type);
+
+    if ( args.size() < 1 ) {
+        zeek::reporter->Error("not enough arguments to Cluster::make_event()");
```

Comment:
`Cluster::make_event()` isn't a thing yet? Also applies to similar comments in the header file. It was probably just too tedious to split cleanly, and that's okay ... but I'm flagging in case it would make sense to introduce more of this in Part 2.

Reply:
> It was probably just too tedious to split cleanly, and that's okay

Too tedious fits: I basically added whole files, accepting that there will be forward references in places. I could see pieces of `cluster.bif` being added here already, but not used.

Reply:
No problem, thanks.